pkg/distributor/query_ingest_storage_test.go

```diff
@@ -0,0 +1,513 @@
+				"zone-b": {numIngesters: 5, happyIngesters: 5},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+					makeWriteRequest(0, 1, 0, false, false, "foo3"),
+					makeWriteRequest(0, 1, 0, false, false, "foo4"),
+				},
+				"zone-b": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+					makeWriteRequest(0, 1, 0, false, false, "foo3"),
+					makeWriteRequest(0, 1, 0, false, false, "foo4"),
+				},
+			},
+			preferZone:               "zone-a",
+			minimizeIngesterRequests: true,
+			matchers:                 []*labels.Matcher{selectAllSeriesMatcher},
+			expectedResponse:         expectedResponse(0, 1, false, "foo0", "foo1", "foo2", "foo3", "foo4"),
+			expectedQueriedIngesters: 4 /* zone-a ingesters (excluding JOINING one) */ + 1, /* zone-b ingester as a fallback for the LEAVING one */
+		},
+		"should fail if all the ingesters owning a partition are in LEAVING state": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {numIngesters: 5, happyIngesters: 5, ringStates: []ring.InstanceState{ring.LEAVING, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE}},
+				"zone-b": {numIngesters: 5, happyIngesters: 5, ringStates: []ring.InstanceState{ring.LEAVING, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE}},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+					makeWriteRequest(0, 1, 0, false, false, "foo3"),
+					makeWriteRequest(0, 1, 0, false, false, "foo4"),
+				},
+				"zone-b": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+					makeWriteRequest(0, 1, 0, false, false, "foo3"),
+					makeWriteRequest(0, 1, 0, false, false, "foo4"),
+				},
+			},
+			preferZone:               "zone-a",
+			minimizeIngesterRequests: true,
+			matchers:                 []*labels.Matcher{selectAllSeriesMatcher},
+			expectedErr:              ring.ErrTooManyUnhealthyInstances,
+		},
+		"should fail if all the ingesters owning a partition are in JOINING state": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {numIngesters: 5, happyIngesters: 5, ringStates: []ring.InstanceState{ring.JOINING, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE}},
+				"zone-b": {numIngesters: 5, happyIngesters: 5, ringStates: []ring.InstanceState{ring.JOINING, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE}},
+			},
```

Comment:
you can also add a case where different ingesters in the two zones are LEAVING (or JOINING) but the query still succeeds

Reply:
Good idea, done in https://github.com/grafana/mimir/pull/7388/commits/19650e3de24bd7bfc6489de00c0ee5c55eee3972